metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,165 +1,337 @@
+      cachedSuites = index.getSuites(buildTarget)
+      (symbol, fullyQualifiedClassName) <- testSymbols
+        .readOnlySnapshot()
+        .toList
+      // IMPORTANT this check is meant to check for class name, not a symbol
+      fullyQualifiedName = FullyQualifiedName(fullyQualifiedClassName)
+      if !cachedSuites.contains(fullyQualifiedName)
+      entryWithDoc <- {
+        val mSymbol = mtags.Symbol(symbol)
+        computeTestEntry(
+          buildTarget,
+          mSymbol,
+          fullyQualifiedName
+        )
+      }
+    } yield entryWithDoc
+    entries.groupBy(_._1.buildTarget)
+  }
+
+  /**
+   * Compute BuildTargetUpdates from added and deleted entries.
+   * For added entry check if it's located in currently opened file, if yes try to find test cases for it.
+   */
+  private def getBuildTargetUpdates(
+      deletedSuites: Map[BuildTarget, Seq[TestExplorerEvent]],
+      addedSuites: Map[BuildTarget, Seq[TestExplorerEvent]],
+      addedTestCases: Map[BuildTarget, Seq[TestExplorerEvent]]
+  ): Seq[BuildTargetUpdate] = {
+
+    val buildTargetUpdates =
+      (deletedSuites.keySet ++ addedSuites.keySet ++ addedTestCases.keySet)
+        .map { buildTarget =>
+          val deleted = deletedSuites.get(buildTarget).getOrElse(Seq.empty)
+          val added = addedSuites.get(buildTarget).getOrElse(Seq.empty)
+          val added0 = addedTestCases.get(buildTarget).getOrElse(Seq.empty)
+          BuildTargetUpdate(buildTarget, deleted ++ added ++ added0)
+        }
+        .filterNot(_.events.isEmpty())
+        .toSeq
+    buildTargetUpdates
+  }
+
+  private def getSemanticDb(
+      symbol: mtags.Symbol
+  ): Option[(mtags.SymbolDefinition, TextDocument)] = {
+    for {
+      definition <- symbolIndex.definition(symbol)
+      doc <- semanticdbs.textDocument(definition.path).documentIncludingStale
+    } yield (definition, doc)
+  }
+
   private def computeTestEntry(
       buildTarget: BuildTarget,
-      symbol: String,
-      fullyQualifiedClassName: String
-  ): Unit = {
-    definitionProvider.toLocation(symbol, List(buildTarget.getId)).foreach {
-      location =>
-        // fullyQualifiedClassName always contains at least one element - class name
-        val fullyQualifiedName = fullyQualifiedClassName.split('.').toList
-
-        val testClass = TestSuiteDiscoveryResult.TestSuite(
-          fullyQualifiedClassName,
-          fullyQualifiedName.takeRight(1).head,
-          location
+      symbol: mtags.Symbol,
+      fullyQualifiedName: FullyQualifiedName
+  ): Option[(TestEntry, TextDocument)] = {
+    // fullyQualifiedClassName always contains at least one element - class name
+    val fullyQualifiedClassName = fullyQualifiedName.value.split('.').toSeq
+    val entryWithDocumentOpt =
+      for {
+        (definition, doc) <- getSemanticDb(symbol)
+        location <- doc.toLocation(definition.path.toURI.toString, symbol.value)
+        className <- fullyQualifiedClassName.takeRight(1).headOption
```

Comment:
`last` should be safe as it can't be empty after `split('.')`
```suggestion
    val className = fullyQualifiedName.value.split('.').last
    val entryWithDocumentOpt =
      for {
        (definition, doc) <- getSemanticDb(symbol)
        location <- doc.toLocation(definition.path.toURI.toString, symbol.value)
```